crates/containerd-shim-wasm/src/sandbox/shim/otel.rs

```diff
@@ -0,0 +1,164 @@
+use opentelemetry_sdk::{runtime, trace as sdktrace};
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt as _;
+use tracing_subscriber::layer::SubscriberExt as _;
+use tracing_subscriber::{EnvFilter, Registry};
+
+const OTEL_EXPORTER_OTLP_PROTOCOL_HTTP_PROTOBUF: &str = "http/protobuf";
+const OTEL_EXPORTER_OTLP_PROTOCOL_GRPC: &str = "grpc";
+
+/// Configuration struct for OpenTelemetry setup.
+pub struct Config {
+    otel_endpoint: String,
+    otel_protocol: String,
+}
+
+/// Initializes a new OpenTelemetry tracer with the OTLP exporter.
+///
+/// Returns a `Result` containing the initialized tracer or a `TraceError` if initialization fails.
+///
+/// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options
+impl Config {
+    /// Creates a new builder for `OtelConfig`.
+    pub fn builder() -> ConfigBuilder {
+        ConfigBuilder::default()
+    }
+
+    /// Initializes the tracer, sets up the telemetry and subscriber layers, and sets the global subscriber.
+    pub fn init(&self) -> anyhow::Result<ShutdownGuard> {
+        let tracer = self.init_tracer()?;
+        let telemetry = tracing_opentelemetry::layer().with_tracer(tracer);
+        set_text_map_propagator(TraceContextPropagator::new());
+
+        let filter = EnvFilter::try_new("info,h2=off")?;
+
+        let subscriber = Registry::default().with(telemetry).with(filter);
+
+        tracing::subscriber::set_global_default(subscriber)?;
+        Ok(ShutdownGuard)
+    }
+
+    pub fn get_trace_context() -> anyhow::Result<String> {
+        // propogate the context
+        let mut injector: HashMap<String, String> = HashMap::new();
+        global::get_text_map_propagator(|propagator| {
+            // retrieve the context from `tracing`
+            propagator.inject_context(&Span::current().context(), &mut injector);
+        });
+        Ok(serde_json::to_string(&injector)?)
+    }
+
+    pub fn set_trace_context(trace_context: &str) -> anyhow::Result<()> {
+        let extractor: HashMap<String, String> = serde_json::from_str(trace_context)?;
+        let context = global::get_text_map_propagator(|propagator| propagator.extract(&extractor));
+        Span::current().set_parent(context);
+        Ok(())
+    }
+
+    fn init_tracer_http_protobuf(&self) -> SpanExporterBuilder {
+        opentelemetry_otlp::new_exporter()
+            .http()
+            .with_endpoint(&self.otel_endpoint)
+            .into()
+    }
+
+    fn init_tracer_grpc(&self) -> SpanExporterBuilder {
+        opentelemetry_otlp::new_exporter()
+            .tonic()
+            .with_endpoint(&self.otel_endpoint)
+            .into()
+    }
+
+    fn init_tracer(&self) -> Result<opentelemetry_sdk::trace::Tracer, TraceError> {
+        let exporter = match self.otel_protocol.as_str() {
+            OTEL_EXPORTER_OTLP_PROTOCOL_HTTP_PROTOBUF => self.init_tracer_http_protobuf(),
```

Comment:
The sdk should already have a default protocol set in the exporter config.
Do we need to have a custom setup for this?
ref: https://docs.rs/opentelemetry-otlp/latest/src/opentelemetry_otlp/exporter/mod.rs.html#81-93

Reply:
The SDK determines the default protocol based on static features, but if we want to support dynamic protocol , which is to parse environment variable to determine which protocol to use, we can't use the `default()` for `ExportConfig`.

Reply:
Ah you are right, I was thinking this was coming from env.
Kinda frustrating the rust sdk isn't setup for this.